File: backend/app/models/slot.py
from sqlmodel import SQLModel, Field, Relationship
from typing import Optional, List
from datetime import date, time, datetime, timezone
from enum import Enum

class SlotStatus(str, Enum):
    AVAILABLE = "available"
    BOOKED = "booked"
    BLOCKED = "blocked"
    IN_COMPETITION = "in_competition"

class Slot(SQLModel, table=True):
    __tablename__ = "slots"
    
    id: Optional[int] = Field(default=None, primary_key=True)
    venue_id: int = Field(foreign_key="venues.id")
    slot_date: date
    start_time: time
    duration: int = Field(default=90)
    base_price: int
    current_price: int
    status: SlotStatus = Field(default=SlotStatus.AVAILABLE)
    is_competition_enabled: bool = Field(default=False)
    competition_winner_id: Optional[int] = Field(default=None, foreign_key="price_competitions.id", sa_column_kwargs={"nullable": True})
    is_contract_slot: bool = Field(default=False)
    contract_id: Optional[int] = Field(foreign_key="contracts.id", default=None)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    
    venue: "Venue" = Relationship(back_populates="slots")
    bookings: List["Booking"] = Relationship(back_populates="slot")
    competitions: List["PriceCompetition"] = Relationship(back_populates="slot", sa_relationship_kwargs={"foreign_keys": "[PriceCompetition.slot_id]"})
    competition_winner: Optional["PriceCompetition"] = Relationship(
        back_populates="winning_slots",
        sa_relationship_kwargs={"foreign_keys": "[Slot.competition_winner_id]"}
    )
    contract_reference: Optional["ContractSlot"] = Relationship(back_populates="slot", )

File: frontend/src/services/slot.ts
import apiClient from './api'

export interface Slot {
  id: number
  venue_id: number
  slot_date: string
  start_time: string
  duration: number
  base_price: number
  current_price: number
  status: 'available' | 'booked' | 'blocked' | 'in_competition'
  is_competition_enabled: boolean
}

export const slotService = {
  // دریافت سانس‌های یک سالن برای یک تاریخ
  getByVenueAndDate: async (venueId: number, date: string): Promise<Slot[]> => {
    const response = await apiClient.get(`/slots/venue/${venueId}`, { params: { slot_date: date } })
    return response.data
  },

  // دریافت سانس‌های آزاد یک سالن برای یک تاریخ
  getAvailableByVenueAndDate: async (venueId: number, date: string): Promise<Slot[]> => {
    const response = await apiClient.get(`/slots/venue/${venueId}/available`, { params: { slot_date: date } })
    return response.data
  },

  // تولید سانس‌ها برای یک روز
  generateForDate: async (venueId: number, date: string): Promise<{ message: string }> => {
    const response = await apiClient.post(`/slots/venue/${venueId}/generate`, null, { params: { slot_date: date } })
    return response.data
  },

  // دریافت سانس‌ها برای بازه تاریخ
  getByVenueAndDateRange: async (venueId: number, startDate: string, endDate: string): Promise<Slot[]> => {
    const response = await apiClient.get(`/slots/venue/${venueId}/range`, { params: { start_date: startDate, end_date: endDate } })
    return response.data
  },
}